pkg/pushsync/pushsync.go

```diff
@@ -352,129 +368,175 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
-		// should not count the retry and try with a different peer again
-		if attempted {
 			allowedRetries--
-		}
-		if err != nil {
-			var timeToSkip time.Duration
-			switch {
-			case errors.Is(err, accounting.ErrOverdraft):
-				skipPeers = append(skipPeers, peer)
-			default:
-				timeToSkip = sanctionWait
+			allowedAttempts--
+
+			peer, err := nextPeer()
+			if err != nil {
+				return nil, err
 			}
 
-			logger.Debugf("pushsync: could not push to peer %s: %v", peer, err)
+			ps.metrics.TotalSendAttempts.Inc()
 
-			// if the node has warmed up AND no other closer peer has been tried
-			if ps.warmedUp() && timeToSkip > 0 {
-				ps.skipList.Add(ch.Address(), peer, timeToSkip)
-				ps.metrics.TotalSkippedPeers.Inc()
-				logger.Debugf("pushsync: adding to skiplist peer %s", peer.String())
+			skipPeers = append(skipPeers, peer)
+
+			ctxd, cancel := context.WithCancel(ctx)
+			// cancel only after defaultTTL to allow pushPeer to fully complete for inflight requests
+			time.AfterFunc(defaultTTL, cancel)
+
+			go ps.pushPeer(ctxd, resultChan, peer, ch, origin)
+
+			// reached the limit, do not set timer to retry
+			if allowedRetries <= 0 || allowedAttempts <= 0 {
+				continue
+			}
+
+			// retry
+			timer.Reset(p90TTL)
+
+		case result := <-resultChan:
+
+			ps.measurePushPeer(result.pushTime, result.err)
+
+			if result.err == nil {
+				return result.receipt, nil
 			}
+
 			ps.metrics.TotalFailedSendAttempts.Inc()
-			if allowedRetries > 0 {
-				continue
+			logger.Debugf("pushsync: could not push to peer %s: %v", result.peer, result.err)
+
+			// peer returned early with error, do not count as an attempt
+			if !result.attempted {
+				allowedAttempts++
 			}
-			return nil, err
-		}
 
-		return r, nil
+			if ps.warmedUp() && !errors.Is(result.err, accounting.ErrOverdraft) {
+				ps.skipList.Add(ch.Address(), result.peer, sanctionWait)
+				ps.metrics.TotalSkippedPeers.Inc()
+				logger.Debugf("pushsync: adding to skiplist peer %s", result.peer.String())
+			}
+
+			if allowedRetries <= 0 || allowedAttempts <= 0 {
+				return nil, ErrNoPush
+			}
+
+			// retry
+			timer.Reset(0)
+		}
 	}
+}
 
-	return nil, ErrNoPush
+func (ps *PushSync) measurePushPeer(t time.Time, err error) {
+	var status string
+	if err != nil {
+		status = "failure"
+	} else {
+		status = "success"
+	}
+	ps.metrics.PushToPeerTime.WithLabelValues(status).Observe(time.Since(t).Seconds())
 }
 
-func (ps *PushSync) pushPeer(ctx context.Context, peer swarm.Address, ch swarm.Chunk, origin bool) (*pb.Receipt, bool, error) {
+func (ps *PushSync) pushPeer(ctx context.Context, resultChan chan<- receiptResult, peer swarm.Address, ch swarm.Chunk, origin bool) {
+
+	var (
+		err       error
+		receipt   pb.Receipt
+		attempted bool
+		now       = time.Now()
+	)
+
+	defer func() {
+		select {
+		case resultChan <- receiptResult{pushTime: now, peer: peer, err: err, attempted: attempted, receipt: &receipt}:
```

Comment:
So I am not able to see where we limit the no of in-flight requests to two. Could it happen that we do more than 2 requests and then lose some results here?

Reply:
We could actually have many inflight requests. It's limited by allowedAttempts.
In the case that a receipt is received from some request, the results from the other requests are ignored.